crates/cast/src/main.rs

```diff
@@ -372,3 +410,59 @@ fn update_cast_config(config: &mut CastConfig, cli: &Cli) {
 
     config.accounts_file = Utf8PathBuf::from(shellexpand::tilde(&new_accounts_file).to_string());
 }
+
+struct PackageData {
+    pub manifest_path: Result<Utf8PathBuf>,
+    pub metadata: Result<Metadata>,
+    pub package: Result<PackageMetadata>,
+}
+
+// Lazily get all the required package information from CLI
+fn get_package_data(cli: &Cli) -> PackageData {
+    let mut result = PackageData {
+        manifest_path: Err(anyhow!("Could not retrieve the manifest path")),
+        metadata: Err(anyhow!("Could not retrieve metadata")),
+        package: Err(anyhow!("Could not retrieve package metadata")),
+    };
+
+    let manifest_path = match cli.path_to_scarb_toml.clone() {
+        Some(path) => Ok(path),
+        None => get_scarb_manifest().context("Failed to obtain manifest path from scarb"),
+    };
+
+    // Make sure the path exists
+    match &manifest_path {
+        Ok(path) => {
+            if path.exists() {
+                result.manifest_path = manifest_path;
+            }
+        }
+        Err(_) => {
+            result.manifest_path = manifest_path;
+        }
+    };
+
+    if result.manifest_path.is_err() {
+        return result;
+    }
+
+    let metadata = get_scarb_metadata_with_deps(result.manifest_path.as_ref().unwrap());
+
+    result.metadata = metadata;
+    if result.metadata.is_err() {
+        return result;
+    }
+
+    let package = match cli
+        .packages_filter
+        .clone()
+        .into_packages_filter()
+        .match_one(result.metadata.as_ref().unwrap())
+    {
+        Ok(data) => Ok(data),
+        Err(_) => get_first_package_from_metadata(result.metadata.as_ref().unwrap()),
```

Comment:
Is this guaranteed by scarb for the main package to be the first in the metadata list?

Reply:
This is the way it worked to this date, so it's not breaking anything, I think it's a sensible strategy

Reply:
Is it agreed that we should always take the config from the first workspace member if there is no package specified? What if `--workspace` was specified (I guess `match_one` fails, right?)? 
cc @THenry14

Reply:
It doesn't look like a good design decision to me. It is basically taking a config of a random workspace member and then e.g. sending transactions using it. In the following scenario:

```
crates/
  A/
    Scarb.toml
  B/
    Scarb.toml
Scarb.toml
```

if I use `sncast` from `crates` directory I basically don't know, as the user, from what crate my config will be taken from. Imo we just shouldn't read the config if it is not precisely specified. By "precisely specified I mean:
- `sncast -p A`
- `sncast --manifest-path crates/A/Scarb.toml` 
- `sncast` called from A/ or A/src etc.

All of the cases should be tested too (prob they are, didn't look into this PR that carefully)

Reply:
the only scenarios when we need to think about scarb and its directory layout are scripts and declare.

For scripts we need to pass package name, which (afaik) should be unique in the setup you've described (please, correct me if I'm wrong).
For declare, we need to provide the contract name - now, I'm not sure about it's uniqueness, this indeed may be problematic in some cases.

Cast allow to pass manifest path to it (`--path-to-scarb-toml`), you can of course always call it from specific dirs.

Given all this, I understand your concern about taking just first package - if I remember correctly this was something suggested to us by scarb team way back when there was no workspace support in scarb. Now that things changed, we should probably indeed take different approach.

I would suggest that (for declare) we accept an argument 'package_name' that is required if there are more packages in workspace than 1. That way we should ensure contracts from proper packages are always taken. Does that sound good?

Reply:
Discussed on the meeting